syncd/syncd.cpp

```diff
@@ -3575,7 +3593,54 @@ int syncd_main(int argc, char **argv)
                  */
 
                 shutdownType = handleRestartQuery(*restartQuery);
-                break;
+                if (shutdownType != SYNCD_RESTART_TYPE_PRE_SHUTDOWN) {
+                    // break out the event handling loop to shutdown syncd
+                    break;
+                }
+
+                // Handle switch pre-shutdown and wait for the final shutdown
+                // event
+
+                SWSS_LOG_TIMER("warm pre-shutdown");
+
+                FlexCounter::removeAllCounters();
+                stopNotificationsProcessingThread();
+
+                sai_attribute_t attr;
+
+                attr.id = SAI_SWITCH_ATTR_RESTART_WARM;
+                attr.value.booldata = true;
+
+                status = sai_switch_api->set_switch_attribute(gSwitchId, &attr);
+
+                if (status != SAI_STATUS_SUCCESS)
+                {
+                    SWSS_LOG_ERROR("Failed to set SAI_SWITCH_ATTR_RESTART_WARM=true: %s for pre-shutdown",
+                            sai_serialize_status(status).c_str());
+                    shutdownType = SYNCD_RESTART_TYPE_COLD;
+                    warmRestartTable->hset("warm-shutdown", "state", "set-flag-failed");
+                    continue;
+                }
+
+                attr.id = SAI_SWITCH_ATTR_PRE_SHUTDOWN;
+                attr.value.booldata = true;
+
+                status = sai_switch_api->set_switch_attribute(gSwitchId, &attr);
+                if (status == SAI_STATUS_SUCCESS)
+                {
+                    warmRestartTable->hset("warm-shutdown", "state", "pre-shutdown-succeeded");
```

Comment:
Once the switch is in pre-shutdown-succeeded state, is there need to prevent syncd from fetching events other than SYNCD_RESTART_TYPE_WARM,  in case there is any pending request data in asicDB?

Reply:
The draining event logic is likely more important in syncd warm restart case. Since we paused orchagent before pre-shutdown, and also stopped event notification from SAI. We have slim chance running into issue in system level warm reboot.

With that said, what you pointed out is a valid issue.

Reply:
Kamil already has a PR outstanding to solve this issue. The code could extend to drain event queue before pre-shutdown.

Since pre-shutdown is only called after pausing orchagent,  adding that should solve the queue draining issue.